deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store_dets.erl

```diff
@@ -2,68 +2,308 @@
-    mqtt_msg() | mqtt_msg_v0() | undefined.
-lookup(Topic, #store_state{table = T}) ->
-  case dets:lookup(T, Topic) of
-    []      -> undefined;
-    [#retained_message{mqtt_msg = Msg}] -> Msg
+    RootId =
+      case find_root_node(NodeTable, EdgeTable) of
+        {ok, Id} ->
+          ?LOG_INFO("Recovered existing RootId: ~p", [Id]),
+          Id;
+        error ->
+          NewId = make_node_id(),
+          ?LOG_INFO("Creating new RootId: ~p", [NewId]),
+          ok = dets:insert(NodeTable, {NewId, 0, false}),
+          NewId
+      end,
+
+    State =
+      #store_state{node_table = NodeTable,
+                   edge_table = EdgeTable,
+                   msg_table = MsgTable,
+                   root_id = RootId,
+                   dir = Dir,
+                   vhost = VHost},
+    {ok, State, Expire}
+  catch
+    error:Reason ->
+      ?LOG_ERROR("Failed to recover MQTT retained message store: ~p", [Reason]),
+      {error, uninitialized}
   end.
 
+-spec insert(topic(), mqtt_msg(), store_state()) -> ok.
+insert(Topic, Msg, #store_state{} = State) ->
+  Words = split_topic(Topic),
+  Insert =
+    fun() ->
+       NodeId = follow_or_create_path(Words, State),
+       % Mark node as topic end and store message
+       update_node(NodeId, true, State),
+       % Replace any existing message for this topic
+       dets:match_delete(State#store_state.msg_table, {NodeId, Topic, '_'}),
+       dets:insert(State#store_state.msg_table, {NodeId, Topic, Msg})
+    end,
+  Insert(),
+  %? TODO: Wrap in transaction since we're modifying multiple tables
+  ok.
+
+-spec lookup(topic(), store_state()) -> [mqtt_msg()] | [mqtt_msg_v0()] | [].
+lookup(Topic, #store_state{} = State) ->
+  Words = split_topic(Topic),
+  Matches = match_pattern_words(Words, State#store_state.root_id, State, []),
+  Values =
+    lists:flatmap(fun(NodeId) ->
+                     case dets:lookup(State#store_state.msg_table, NodeId) of
+                       [{_NodeId, _Topic, Value} | _] -> [Value];
+                       [] -> []
+                     end
+                  end,
+                  Matches),
+  Values.
+
 -spec delete(topic(), store_state()) -> ok.
-delete(Topic, #store_state{table = T}) ->
-  ok = dets:delete(T, Topic).
+delete(Topic, State) ->
+  Words = split_topic(Topic),
+  % Wrap in transaction since we're modifying multiple tables
+  Delete =
+    fun() ->
+       case follow_path(Words, State) of
+         {ok, NodeId} ->
+           dets:match_delete(State#store_state.msg_table, {NodeId, Topic, '_'}),
+           case dets:lookup(State#store_state.msg_table, NodeId) of
+             [] ->
+               update_node(NodeId, false, State),
+               maybe_clean_path(NodeId, State);
+             _ -> ok
+           end;
+         error -> ok
+       end
+    end,
+  Delete(),
+  %? TODO: Wrap in transaction since we're modifying multiple tables
+  % rabbit_mnesia:execute_mnesia_transaction(Delete),
+  ok.
```

Comment:
Should this operation be worked on to be atomic?

Reply:
Same as above.